### PR TITLE
Implement a constructor for nested_tensor that is similar to torch.tensor()

### PR DESCRIPTION
**Pull request by ani300**
Summary: This diff merges both previous implementations of constructors for nested tensors, the one from lists of tensors and the one with arbitrary python lists, adn implements it in pytorch core so no extensions are needed to construct NT.

### CODE REVIEW

File: test/test_nestedtensor.py [hunk 1]
```diff
@@ -93,6 +97,79 @@ def random_nt(device, dtype, num_tensors, max_dims, min_dims=None):
...
+        data = []
+        nested_tensor_ref_list = []
+        for _ in range(batch_size):
+            if max_seq_len == 0:
+                length = 0
+            else:
+                length = np.random.randint(low=1, high=max_seq_len)
+            row = list(np.random.randint(low=0, high=VOCAB_SIZE, size=(length,)))
+            data.append(row)
+            nested_tensor_ref_list.append(torch.Tensor(row))
```
Comment (cpuhrsch) [inline test/test_nestedtensor.py:115]: nit: Please us `torch.tensor`

File: test/test_nestedtensor.py [hunk 2]
```diff
@@ -93,6 +97,79 @@ def random_nt(device, dtype, num_tensors, max_dims, min_dims=None):
...
+        nested_tensor = torch.nested.nested_tensor(data, dtype=torch.int)
+        nested_tensor_list = nested_tensor.unbind()
+        for id in range(batch_size):
+            assert torch.equal(
```
Comment (cpuhrsch) [inline test/test_nestedtensor.py:119]: Please use self.assertEqual

File: test/test_nestedtensor.py [hunk 3]
```diff
@@ -93,6 +97,79 @@ def random_nt(device, dtype, num_tensors, max_dims, min_dims=None):
...
+                nested_tensor_list[id], nested_tensor_ref_list[id].type(torch.int64)
+            )
+
+    @given(
+        batch_size=st.integers(min_value=2, max_value=4),
+        max_seq_len=st.integers(min_value=2, max_value=4),
+    )
+    @settings(deadline=None)
+    def test_3d_nested_tensor(self, batch_size: int, max_seq_len: int):
```
Comment (cpuhrsch) [inline test/test_nestedtensor.py:128]: Above comments also apply to the tests below

File: torch/csrc/api/include/torch/nested.h [hunk 4]
```diff
@@ -1,7 +1,10 @@
 #pragma once
 
 #include <ATen/ATen.h>
-#include <torch/torch.h>
+#include <torch/csrc/api/include/torch/detail/TensorDataContainer.h>
+#include <algorithm>
+#include <stdexcept>
+#include "ATen/core/ATen_fwd.h"
```
Comment (cpuhrsch) [inline torch/csrc/api/include/torch/nested.h:7]: Please use angle brackets

File: torch/csrc/api/include/torch/nested.h [hunk 5]
```diff
@@ -12,22 +15,42 @@ namespace nested {
...
+}
+
+inline at::Tensor nested_tensor(
+    at::ArrayRef<detail::TensorDataContainer> nested_tensor_data,
+    const at::TensorOptions& options = {}) {
+  if (nested_tensor_data.size() > 0) {
+    if (nested_tensor_data[0].is_init_list()) {
+      // TODO: Implement conversion from InitList to NT for C++,
+      // link into existing infrastructure at torch/csrc/utils/nested.cpp
+      throw std::runtime_error(
```
Comment (cpuhrsch) [inline torch/csrc/api/include/torch/nested.h:45]: Please use `TORCH_CHECK`

File: torch/csrc/api/include/torch/nested.h [hunk 6]
```diff
@@ -36,10 +59,10 @@ inline Tensor nested_tensor(
 /// https://pytorch.org/docs/master/nested.html#torch.nested.as_nested_tensor
 ///
 /// ```
-inline Tensor as_nested_tensor(
-    TensorList list,
-    c10::optional<ScalarType> dtype = c10::nullopt,
-    c10::optional<Device> device = c10::nullopt) {
+inline at::Tensor as_nested_tensor(
```
Comment (cpuhrsch) [inline torch/csrc/api/include/torch/nested.h:62]: Why are these changes needed?

File: torch/csrc/utils/nested.cpp [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+#include <ATen/NestedTensorImpl.h>
+#include <c10/core/ScalarType.h>
+
+#include <stdexcept>
+#include <vector>
+
+namespace torch {
+namespace utils {
+
+long compute_total_size_and_sizes_tensor(
```
Comment (cpuhrsch) [inline torch/csrc/utils/nested.cpp:15]: Is it possible to rely on `_nested_tensor_from_tensor_list` for all the dtype, device, etc. checks?

File: torch/csrc/api/include/torch/nested.h (hunk 6 above)
Comment (ani300) [reply torch/csrc/api/include/torch/nested.h:62]: The compiler started complaining for some reason.

File: torch/csrc/utils/nested.cpp (hunk 7 above)
Comment (ani300) [reply torch/csrc/utils/nested.cpp:15]: Looking at the implementation for both torch::tensor() and _nested_tensor_from_tensor_list, we do the same amount of checks

Comment (cpuhrsch) [reply torch/csrc/utils/nested.cpp:15]: Sure, but this could deduplicate code

File: torch/csrc/api/include/torch/nested.h [hunk 8]
```diff
@@ -12,34 +15,71 @@ namespace nested {
...
     out.requires_grad_(true);
   }
   return out;
 }
 
+// TODO: Implement a fast path for TensorDataContainer
+inline at::Tensor nested_tensor(
+    at::ArrayRef<detail::TensorDataContainer> nested_tensor_data,
+    const at::TensorOptions& options = {}) {
+  if (nested_tensor_data.size() > 0) {
```
Comment (cpuhrsch) [inline torch/csrc/api/include/torch/nested.h:43]: nit: I don't think you should need this check
```
>>> torch.nested.nested_tensor([])
nested_tensor([

])
```

Comment (ani300) [reply torch/csrc/api/include/torch/nested.h:43]: this is for the C++ path, given that python now takes python vars it needs different impls (or at least torch::tensor does so)

File: torch/nested/__init__.py [hunk 9]
```diff
@@ -137,4 +113,34 @@ def as_nested_tensor(tensor_list: List[Tensor], dtype: Optional[DType] = None, d
     >>> pt_small = torch.nested.to_padded_tensor(nt, 2.0, (2, 2, 2))
     RuntimeError: Value in output_size is less than NestedTensor padded size. Truncation is not supported.
 
-""")
+""",
+)
+
+nested_tensor = _add_docstr(
+    _nested.nested_tensor,
+    r"""
```
Comment (mikaylagawarecki) [inline torch/nested/__init__.py:121]: nit: when using `_add_docstr`, in order to render the arguments for the function a line containing the signature of the function needs to be added here (see `to_padded_tensor` above for an example)

(the preview for the docs is available under helpful links [in the pytorch-bot comment on this PR](https://github.com/pytorch/pytorch/pull/88213#issuecomment-1298866390) if you would like to look at the rendering!)

**Review by cpuhrsch — APPROVED** [review]
